src/gc.h

```diff
@@ -0,0 +1,355 @@
+// This file is a part of Julia. License is MIT: http://julialang.org/license
+
+/*
+  allocation and garbage collection
+  . non-moving, precise mark and sweep collector
+  . pool-allocates small objects, keeps big objects on a simple list
+*/
+
+#ifndef JULIA_GC_H
+#define JULIA_GC_H
+
+#include <stdlib.h>
+#include <string.h>
+#ifndef _MSC_VER
+#include <strings.h>
+#endif
+#include <assert.h>
+#include <inttypes.h>
+#include "julia.h"
+#include "julia_internal.h"
+#include "threading.h"
+#ifndef _OS_WINDOWS_
+#include <sys/mman.h>
+#if defined(_OS_DARWIN_) && !defined(MAP_ANONYMOUS)
+#define MAP_ANONYMOUS MAP_ANON
+#endif
+#endif
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+// manipulating mark bits
+
+#define GC_CLEAN 0 // freshly allocated
+#define GC_MARKED 1 // reachable and old
+#define GC_QUEUED 2 // if it is reachable it will be marked as old
+#define GC_MARKED_NOESC (GC_MARKED | GC_QUEUED) // reachable and young
+
+#define GC_PAGE_LG2 14 // log2(size of a page)
+#define GC_PAGE_SZ (1 << GC_PAGE_LG2) // 16k
+#define GC_PAGE_OFFSET (JL_SMALL_BYTE_ALIGNMENT - (sizeof_jl_taggedvalue_t % JL_SMALL_BYTE_ALIGNMENT))
+
+// 8G * 32768 = 2^48
+// It's really unlikely that we'll actually allocate that much though...
+#define REGION_COUNT 32768
+
+#define jl_buff_tag ((uintptr_t)0x4eade800)
+#define jl_malloc_tag ((void*)0xdeadaa01)
+#define jl_singleton_tag ((void*)0xdeadaa02)
+
+// Used by GC_DEBUG_ENV
+typedef struct {
+    uint64_t num;
+    uint64_t next;
+
+    uint64_t min;
+    uint64_t interv;
+    uint64_t max;
+    unsigned short random[3];
+} jl_alloc_num_t;
+
+typedef struct {
+    int sweep_mask;
+    int wait_for_debugger;
+    jl_alloc_num_t pool;
+    jl_alloc_num_t other;
+    jl_alloc_num_t print;
+} jl_gc_debug_env_t;
+
+// This struct must be kept in sync with the Julia type of the same name in base/util.jl
+typedef struct {
+    int64_t     allocd;
+    int64_t     freed;
+    uint64_t    malloc;
+    uint64_t    realloc;
+    uint64_t    poolalloc;
+    uint64_t    bigalloc;
+    uint64_t    freecall;
+    uint64_t    total_time;
+    uint64_t    total_allocd;
+    uint64_t    since_sweep;
+    size_t      interval;
+    int         pause;
+    int         full_sweep;
+} jl_gc_num_t;
+
+// layout for small (<2k) objects
+
+typedef struct _buff_t {
+    union {
+        uintptr_t header;
+        struct _buff_t *next;
+        uintptr_t flags;
+        jl_value_t *type; // 16-bytes aligned
+        struct {
+            uintptr_t gc_bits:2;
+            uintptr_t pooled:1;
+        };
+    };
+    // Work around a bug affecting gcc up to (at least) version 4.4.7
+    // https://gcc.gnu.org/bugzilla/show_bug.cgi?id=36839
+#if !defined(_COMPILER_MICROSOFT_)
+    int _dummy[0];
+#endif
```

Comment:
The comment says the issue exists on 4.4(.7) it's a little unclear which version fixes the issue though since the only comment after 2009 was a comment from 2014 (after GCC 4.9.0 is released) saying the issue is fixed.